### PR TITLE
Dev/anl/web android merge

### PR DESCRIPTION
**Pull request by benanil**

### CODE REVIEW

File: Dependency/ImGui/CMakeCache.txt [hunk 1]
```diff
@@ -0,0 +1,328 @@
+# This is the CMakeCache file.
```
Comment (afraidofdark) [inline Dependency/ImGui/CMakeCache.txt:1]: did you forget this temp file ?

File: Template/Android/app/src/main/cpp/Game.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    material->m_fragmentShader = shader;
+    material->Init();
+
+    cubeMaterial  = std::make_shared<Material>();
+    cubeMaterial->m_vertexShader   = GetShaderManager()->Create<Shader>(ShaderPath("defaultVertex.shader", true));
+    cubeMaterial->m_fragmentShader = GetShaderManager()->Create<Shader>(ShaderPath("defaultFragment.shader", true));
+    cubeMaterial->Init();
+    camera = std::make_shared<Camera>();
+
+    frameBuffer   = std::make_shared<Framebuffer>();
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/Game.cpp:54]: this must be MakeNewPtr

File: Template/Android/app/src/main/cpp/Game.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#pragma once
+
+#include "ToolKit.h"
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/Game.h:29]: these files must be in the common and there should not be any code, sample etc ..

File: Template/Android/app/src/main/cpp/ToolKitMain.cpp [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+
+#include "ToolKit.h"
+#include "Logger.h"
+#include "Game.h"
+#include "Utility.h"
+#include "SceneRenderer.h"
+
+#define CHECK_GL_ERROR()
+
+static void checkGLError(const char* file, int line)
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ToolKitMain.cpp:20]: the naming for the file is wrong. This should be android_main.cpp / h

File: Template/Android/app/src/main/cpp/ToolKitMain.cpp [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+
+      void ContentResize(float width, float height)
+      {
+          OnResizeContentArea(width, height);
+      }
+
+      void Update(float dt) override { }
+  };
+
+  static GameViewport* gameViewport;
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ToolKitMain.cpp:62]: static globals start with g_

File: Template/Android/app/src/main/cpp/Utility.h [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#ifndef ANDROIDGLINVESTIGATIONS_UTILITY_H
+#define ANDROIDGLINVESTIGATIONS_UTILITY_H
+
+#include <cassert>
+#include <android/log.h>
+
+#define TK_LOG(format, ...) __android_log_print(ANDROID_LOG_DEBUG, "TK_LOG", format, ##__VA_ARGS__)
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/Utility.h:7]: You can't do this, pass a lamda to logger that logs to android like we do.

File: Template/Android/app/src/main/cpp/main.cpp [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+#include "ToolKit.h"
+#include "Logger.h"
+#include "ToolKitMain.h"
+
+#include <game-activity/GameActivity.cpp>
+#include <game-text-input/gametextinput.cpp>
+
+extern "C" {
+
+#include <game-activity/native_app_glue/android_native_app_glue.c>
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/main.cpp:13]: why not these thins aren't under ToolKit/Dependencies/Android ???

File: ToolKit/DeferredPass.cpp [hunk 8]
```diff
@@ -98,10 +98,10 @@ namespace ToolKit
     // Set gbuffer
     // 9: Position, 10: Normal, 11: Color, 12: emissive, 14: metallic-roughness,
     // 16: ibl contribution
-    using FAttachment           = Framebuffer::Attachment;
+    using FAttachment = Framebuffer::Attachment;
```
Comment (afraidofdark) [inline ToolKit/DeferredPass.cpp:101]: format file

File: Template/Android/app/src/main/cpp/ToolKitMain.cpp [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+    return static_cast<float>(time_span.count() * 1000.0);
+  }
+
+  void AndroidDevice::ToolKitFrame()
+  {
+    static Timing timing{};
+    timing.CurrentTime = GetMilliSeconds();
+    timing.DeltaTime = timing.CurrentTime - timing.LastTime;
+
+    m_game->Frame(0.01f, gameViewport);
```
Comment (Osman-Fatih-Cakir) [inline Template/Android/app/src/main/cpp/ToolKitMain.cpp:303]: I think we should give timing.DeltaTime as argument instead of 0.01f.

File: ToolKit/MathUtil.cpp [hunk 10]
```diff
@@ -395,7 +396,7 @@ namespace ToolKit
     float closestPickedDistance = FLT_MAX;
     bool hit                    = false;
 
-#ifndef __EMSCRIPTEN__
+#ifndef __clang__
```
Comment (Osman-Fatih-Cakir) [inline ToolKit/MathUtil.cpp:399]: Emscripten uses clang too so this is probably not a problem, but we should check if web build works fine.

File: ToolKit/Texture.h [hunk 11]
```diff
@@ -30,6 +30,8 @@
 #include "ResourceManager.h"
 #include "Types.h"
 
+#define TK_LUT_TEXTURE "GLOBAL_BRDF_LUT_TEXTURE"
```
Comment (Osman-Fatih-Cakir) [inline ToolKit/Texture.h:33]: Name of the texture should mention that this is BRDF look up texture so I think the naming can be TK_BRDF_LUT_TEXTURE.

File: ToolKit/Logger.h [hunk 12]
```diff
@@ -40,15 +40,15 @@ namespace ToolKit
     Command
   };
 
-  typedef std::function<void(LogType, const String&)> ConsoleOutputFn;
+  typedef std::function<void(LogType, const std::string&)> ConsoleOutputFn;
```
Comment (afraidofdark) [inline ToolKit/Logger.h:43]: String&

File: ToolKit/Logger.h [hunk 13]
```diff
@@ -40,15 +40,15 @@ namespace ToolKit
...
   typedef std::function<void()> ClearConsoleFn;
 
   class TK_API Logger
   {
    public:
     Logger();
     ~Logger();
-    void Log(const String& message);
+    void Log(const std::string& message);
```
Comment (afraidofdark) [inline ToolKit/Logger.h:51]: String&

File: ToolKit/Material.cpp [hunk 14]
```diff
@@ -53,9 +54,9 @@ namespace ToolKit
     if (!m_loaded)
     {
       m_loaded = true;
-      ParseDocument("material");
+        ParseDocument("material");
+      }
```
Comment (afraidofdark) [inline ToolKit/Material.cpp:58]: indentation is broken

File: ToolKit/RenderSystem.cpp [hunk 15]
```diff
@@ -111,6 +112,8 @@ namespace ToolKit
     }
   }
 
+  Renderer* RenderSystem::GetRenderer() { return m_renderer; }
```
Comment (afraidofdark) [inline ToolKit/RenderSystem.cpp:115]: You can't expose the renderer. Remove this and change your logic.

File: ToolKit/RenderSystem.h [hunk 16]
```diff
@@ -80,6 +80,7 @@ namespace ToolKit
     void AddRenderTask(RenderTask task);
     void ExecuteRenderTasks();
     void FlushRenderTasks();
+    Renderer* GetRenderer();
```
Comment (afraidofdark) [inline ToolKit/RenderSystem.h:83]: Remove this

File: ToolKit/Renderer.cpp [hunk 17]
```diff
@@ -167,7 +168,7 @@ namespace ToolKit
     SetProjectViewModel(job.WorldTransform, cam);
     m_lights = lights;
     m_cam    = cam;
-    job.Mesh->Init();
+     job.Mesh->Init();
```
Comment (afraidofdark) [inline ToolKit/Renderer.cpp:171]: fix formating.

File: ToolKit/Renderer.cpp [hunk 18]
```diff
@@ -356,9 +357,9 @@ namespace ToolKit
     }
   }
 
-  void Renderer::SetFramebuffer(FramebufferPtr fb, bool clear, const Vec4& color)
+  void Renderer::SetFramebuffer(FramebufferPtr fb, bool clear, const Vec4& color, bool force)
   {
-    if (fb != m_framebuffer)
+    if (force || fb != m_framebuffer)
```
Comment (afraidofdark) [inline ToolKit/Renderer.cpp:362]: you can't make this

File: ToolKit/Renderer.h [hunk 19]
```diff
@@ -70,8 +72,8 @@ namespace ToolKit
     void SetRenderState(const RenderState* const state);
 
     void SetStencilOperation(StencilOperation op);
-    void SetFramebuffer(FramebufferPtr fb, bool clear, const Vec4& color);
-    void SetFramebuffer(FramebufferPtr fb, bool clear = true);
+    void SetFramebuffer(FramebufferPtr fb, bool clear, const Vec4& color, bool force = false);
+    void SetFramebuffer(FramebufferPtr fb, bool clear = true, bool force = false);
```
Comment (afraidofdark) [inline ToolKit/Renderer.h:76]: remove force

File: ToolKit/ResourceManager.cpp [hunk 20]
```diff
@@ -75,6 +75,10 @@ namespace ToolKit
 
   bool ResourceManager::Exist(const String& file) { return m_storage.find(file) != m_storage.end(); }
 
+  template std::shared_ptr<Material> ResourceManager::Copy(ResourcePtr source);
```
Comment (afraidofdark) [inline ToolKit/ResourceManager.cpp:78]: these are not needed clear template instantiations.

File: ToolKit/GlErrorReporter.cpp [hunk 21]
```diff
@@ -25,24 +25,24 @@
...
 namespace ToolKit
 {
 
-  GlReportCallback GlErrorReporter::Report = [](const std::string& msg) -> void { GetLogger()->Log(msg); };
+  GlReportCallback GlErrorReporter::Report = [](const std::string& msg) -> void {  };
 
   void InitGLErrorReport(GlReportCallback callback)
   {
-#ifdef TK_GL_CORE_3_2
+#if defined(TK_GL_CORE_3_2) || defined(__ANDROID__)
```
[the remark this answers is not in the document]
Comment (afraidofdark) [reply ToolKit/GlErrorReporter.cpp:38]: the problem is that, glDebugMessageCallback is not present int webgl, maybe an extention helps

File: ToolKit/Texture.cpp [hunk 22]
```diff
@@ -181,8 +181,8 @@ namespace ToolKit
 
   void Texture::Clear()
   {
-    stbi_image_free(m_image);
-    stbi_image_free(m_imagef);
+    free(m_image);
```
Comment (afraidofdark) [inline ToolKit/Texture.cpp:184]: why not use stbi_image_free ? this will potentially cause memory leaks

File: ToolKit/FileManager.cpp [hunk 23]
```diff
@@ -98,7 +123,10 @@ namespace ToolKit
       }
       else if (fileType == FileType::ImageFloat)
       {
-        return ReadHdriFileFromZip(m_zfile, relativePath, fileInfo);
+        stbi_set_flip_vertically_on_load(true);
```
Comment (afraidofdark) [inline ToolKit/FileManager.cpp:126]: @Osman-Fatih-Cakir take a look at here

File: ToolKit/Texture.cpp [hunk 24]
```diff
@@ -397,9 +397,7 @@ namespace ToolKit
     }
 
     // Load hdri image
-    stbi_set_flip_vertically_on_load(true);
```
Comment (afraidofdark) [inline ToolKit/Texture.cpp:400]: @Osman-Fatih-Cakir Anil converts all floating texture readings.

File: ToolKit/ToolKit.cpp [hunk 25]
```diff
@@ -315,7 +313,11 @@ namespace ToolKit
 
   String DefaultPath()
   {
+#ifdef __ANDROID__
+    static const String res = ConcatPaths({ Main::GetInstance()->m_resourceRoot, "Resources", "Engine"});
```
Comment (afraidofdark) [inline ToolKit/ToolKit.cpp:317]: fix formating and make sure you have formated all toolkit files that you have changed.

File: ToolKit/Types.h [hunk 26]
```diff
@@ -31,8 +31,49 @@
...
+// defined in ToolKitMain.cpp
+extern struct android_app* g_android_app;
+extern struct AAssetManager* g_asset_manager;
+
+#else
+#define TK_LOG(format, ...) GetLogger()->WriteConsole(LogType::Memo, format, ##__VA_ARGS__)
+#endif
+
+// GLM
+#define GLM_FORCE_XYZW_ONLY
```
Comment (afraidofdark) [inline ToolKit/Types.h:50]: you can not put them here

File: ToolKit/Types.h [hunk 27]
```diff
@@ -31,8 +31,49 @@
...
+#include "game-activity/GameActivity.h"
+#include <android/asset_manager.h>
+#include <game-activity/native_app_glue/android_native_app_glue.h>
+
+// defined in ToolKitMain.cpp
+extern struct android_app* g_android_app;
+extern struct AAssetManager* g_asset_manager;
+
+#else
+#define TK_LOG(format, ...) GetLogger()->WriteConsole(LogType::Memo, format, ##__VA_ARGS__)
```
Comment (afraidofdark) [inline ToolKit/Types.h:46]: you have another decleration of this in android template.

File: ToolKit/Types.h [hunk 28]
```diff
@@ -31,8 +31,49 @@
  * relatled structures.
  */
 
+#ifdef __ANDROID__
+#define TK_GL_ES_3_0
+#define TK_DLL_EXPORT
```
Comment (afraidofdark) [inline ToolKit/Types.h:36]: move these declerations to cmake

File: ToolKit/Types.h [hunk 29]
```diff
@@ -44,6 +85,12 @@
 #include <utility>
 #include <vector>
 
+template<class _Tp, class _Up>
```
Comment (afraidofdark) [inline ToolKit/Types.h:88]: move this to utils.h at the bottom std functions section.

File: ToolKit/FileManager.cpp (hunk 23 above)
Comment (Osman-Fatih-Cakir) [reply ToolKit/FileManager.cpp:126]: stbi_image is only included for FileManager and this function is called there. HDRI's will behave same with this change.

File: ToolKit/UIManager.cpp [hunk 30]
```diff
@@ -90,17 +90,19 @@ namespace ToolKit
...
       return me->m_action == EventAction::LeftClick;
     }
-
+#endif
     return false;
   }
 
   bool UIManager::CheckMouseOver(Surface* surface, Event* e, Viewport* vp)
   {
+#ifndef __ANDROID__
```
Comment (afraidofdark) [inline ToolKit/UIManager.cpp:105]: why disable ?

File: ToolKit/Util.cpp [hunk 31]
```diff
@@ -388,7 +388,14 @@ namespace ToolKit
     return false;
   }
 
-  bool HasToolKitRoot(const String& path) { return StartsWith(path, "ToolKit\\") || StartsWith(path, "ToolKit/"); }
+  bool HasToolKitRoot(const String& path) 
+  {
+#ifdef __ANDROID__
+    return false;
```
Comment (afraidofdark) [inline ToolKit/Util.cpp:394]: Do you know why this is used ? Resource may have ToolKit in them returning false will cause errors with default assets.

File: ToolKit/Util.cpp [hunk 32]
```diff
@@ -481,8 +488,6 @@ namespace ToolKit
       return SCENE;
     }
 
-    assert(false && "Resource type does not have a corresponding extension.");
```
Comment (afraidofdark) [inline ToolKit/Util.cpp:484]: why did you remove this ?

File: ToolKit/glm/CMakeLists.txt [hunk 33]
```diff
@@ -0,0 +1,67 @@
+file(GLOB ROOT_SOURCE *.cpp)
```
Comment (afraidofdark) [inline ToolKit/glm/CMakeLists.txt:1]: what is this ?

File: ToolKit/stdafx.h [hunk 34]
```diff
@@ -58,18 +58,15 @@
 #include "glm/gtx/string_cast.hpp"
 #include "glm/gtx/vector_query.hpp"
 
-// Stb
-#include "stb/stb_image.h"
```
Comment (afraidofdark) [inline ToolKit/stdafx.h:62]: why did you remove stb image ?

File: Template/Android/app/src/main/cpp/ToolKitMain.cpp [hunk 35]
```diff
@@ -0,0 +1,457 @@
...
+    timing.TimeAccum += timing.DeltaTime;
+    if (timing.TimeAccum >= 1000.0f)
+    {
+      timing.TimeAccum = timing.FrameCount = 0;
+    }
+
+    timing.LastTime = timing.CurrentTime;
+  }
+
+  inline bool FileHasExtension(const char* path, int size, const char* extension)
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ToolKitMain.cpp:316]: use util.h DecomposePath

File: Template/Android/app/src/main/cpp/ToolKitMain.cpp [hunk 36]
```diff
@@ -0,0 +1,457 @@
...
+      return true;
+  }
+
+  // copys all of the engine assets to internal data folder if already not coppied
+  void AndroidDevice::CopyAllAssetsToDataPath()
+  {
+    String internalDataPath = g_android_app->activity->internalDataPath;
+    using namespace std::filesystem;
+    
+    if (!exists(internalDataPath + "/Resources"))
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ToolKitMain.cpp:339]: all cruly braces must be there {} even for one lines.

File: Template/Android/app/src/main/cpp/CMakeLists.txt [hunk 37]
```diff
@@ -0,0 +1,95 @@
...
+    if(NOT EXISTS ${TOOLKIT_DIR}/ToolKit/ToolKit.h)
+        message(SEND_ERROR "Can't find ToolKit.h in ${TOOLKIT_DIR}")
+    endif()
+endif()
+
+# copy engine resources to Android/assets folder
+if (NOT EXISTS "${CMAKE_SOURCE_DIR}/../assets/Resources")
+    message("copying engine resources to android assets folder")
+
+    file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/../assets/Resources")
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/CMakeLists.txt:34]: our packing code moves all the engine assets in to resource folder of the project and creates a pack file. We need to move the pack file to the android's storage. So this step is not needed. Packing will take care of this.

File: Template/Android/app/src/main/cpp/ToolKitMain.cpp [hunk 38]
```diff
@@ -0,0 +1,457 @@
...
+
+      TK_LOG("Found: %i configs", numConfigs);
+      // create the proper window surface
+      EGLint format;
+      eglGetConfigAttrib(display, config, EGL_NATIVE_VISUAL_ID, &format);
+      EGLSurface surface = eglCreateWindowSurface(display, config, g_android_app->window, nullptr);
+
+      // Create a GLES 3 context
+      const EGLint contextAttribs[] = { EGL_CONTEXT_MAJOR_VERSION, 3,
+                                        EGL_CONTEXT_MINOR_VERSION, 1,
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ToolKitMain.cpp:113]: make this 3_0 or 3_2 based on if defs TK_GL_CORE_3_2 / TK_GL_ES_3_0

File: Template/Android/app/src/main/cpp/ToolKitMain.h [hunk 39]
```diff
@@ -0,0 +1,48 @@
+//
+// Created by Administrator on 8/30/2023.
+//
+
+#ifndef TOOLKITGAME_TOOLKITMAIN_H
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ToolKitMain.h:5]: use pragma once

File: Template/Android/app/src/main/cpp/ToolKitMain.h [hunk 40]
```diff
@@ -0,0 +1,48 @@
...
+typedef void* EGLContext;
+typedef void* EGLDisplay;
+typedef int EGLint;
+typedef unsigned int GLuint;
+
+namespace ToolKit 
+{
+  class Main; 
+
+  class AndroidDevice
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ToolKitMain.h:20]: we should have a coreapp class and all the platforms and apps must extend that.

File: Template/Android/app/src/main/cpp/Utility.h (hunk 6 above)
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/Utility.h:7]: while refactoring the TK_LOG to AndroidLog move this to cpp where it used and remove the file.

move tk_log to types.h which calls GetLogger()->WriteConsole
and move AndroidLog __android_log_print to main.cpp and set it as the logger callback.

File: Template/Android/app/src/main/cpp/ioapi.h [hunk 41]
```diff
@@ -0,0 +1,210 @@
+/* ioapi.h -- IO base function header for compress/uncompress .zip
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/ioapi.h:1]: move this to Dependencies/Android

File: Template/Android/app/src/main/cpp/unzip/crypt.h [hunk 42]
```diff
@@ -0,0 +1,131 @@
+/* crypt.h -- base code for crypt/uncrypt ZIPfile
```
Comment (afraidofdark) [inline Template/Android/app/src/main/cpp/unzip/crypt.h:1]: move this to dependencies/android

File: Template/Android/app/src/main/java/com/otyazilim/myapplication/MainActivity.java [hunk 43]
```diff
@@ -0,0 +1,32 @@
+package com.otyazilim.myapplication;
+
+import android.view.View;
+
+import com.google.androidgamesdk.GameActivity;
+
+public class MainActivity extends GameActivity {
+    static {
+        System.loadLibrary("myapplication");
```
Comment (afraidofdark) [inline Template/Android/app/src/main/java/com/otyazilim/myapplication/MainActivity.java:9]: use a better name

File: Template/Android/app/src/main/res/drawable/ic_launcher_background.xml [hunk 44]
```diff
@@ -0,0 +1,170 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (afraidofdark) [inline Template/Android/app/src/main/res/drawable/ic_launcher_background.xml:1]: make our own icon

Comment (afraidofdark) [reply Template/Android/app/src/main/res/drawable/ic_launcher_background.xml:1]: our shipment must provide a menu for the icons

File: Template/Bin/shell_minimal.html
Comment (afraidofdark) [inline Template/Bin/shell_minimal.html:1]: don't remove

File: Template/CMakeHotReload.cmake
Comment (afraidofdark) [inline Template/CMakeHotReload.cmake:1]: don't remove

File: Template/Template.vcxproj
Comment (afraidofdark) [inline Template/Template.vcxproj:1]: don't remove

File: Template/Template.vcxproj.filters
Comment (afraidofdark) [inline Template/Template.vcxproj.filters:1]: don't remove

File: Template/Template.vcxproj.user
Comment (afraidofdark) [inline Template/Template.vcxproj.user:1]: don't remove

File: Template/config/Engine.settings
Comment (afraidofdark) [inline Template/config/Engine.settings:1]: don't remove and make the folder upper case Config

File: ToolKit/FileManager.cpp [hunk 45]
```diff
@@ -31,14 +31,39 @@
...
+#ifdef __ARM_FP
+# define STBI_NEON
+#endif
 #include "stb/stb_image.h"
 
-#include "DebugNew.h"
-
 namespace ToolKit
 {
+  File File::Open(StringView path, StringView mode)
```
Comment (afraidofdark) [inline ToolKit/FileManager.cpp:45]: remove unused codes.

File: ToolKit/FileManager.cpp [hunk 46]
```diff
@@ -418,23 +449,24 @@ namespace ToolKit
...
+      File::Close(f);
       zipCloseFileInZip(zfile);
       return false;
     }
 
     char* fileData = reinterpret_cast<char*>(malloc((flen + 1) * static_cast<uint>(sizeof(char))));
-    red            = fread(fileData, flen, 1, f);
-    ret            = zipWriteInFileInZip(zfile, fileData, static_cast<uint>(red * flen));
+    red            = f.Read(fileData, flen);
+    ret            = zipWriteInFileInZip(zfile, fileData, static_cast<uint>(red));
```
Comment (afraidofdark) [inline ToolKit/FileManager.cpp:459]: while getting bakc the changes make sure this line stay as is. Because there is a fix here
no "red * flen"

File: ToolKit/FileManager.cpp [hunk 47]
```diff
@@ -459,7 +491,7 @@ namespace ToolKit
   {
     String extrFilesPathStr   = ConcatPaths({path, "..", "ExtraFiles.txt"});
     const char* extrFilesPath = extrFilesPathStr.c_str();
-    if (!std::filesystem::exists(extrFilesPath))
+    if (!CheckSystemFile(extrFilesPath))
```
Comment (afraidofdark) [inline ToolKit/FileManager.cpp:494]: also preseve CheckSystemFile instead of std::filesystem::exist

File: ToolKit/FileManager.h [hunk 48]
```diff
@@ -28,11 +28,30 @@
 
 #include "Types.h"
 
+#include <variant>
+#include <zlib.h>
+
+#ifdef __ANDROID__
+#include "../unzip/unzip.h"
```
Comment (afraidofdark) [inline ToolKit/FileManager.h:35]: change these lines after moving android files. Android/Zlib/unzip.h

File: ToolKit/FileManager.h [hunk 49]
```diff
@@ -28,11 +28,30 @@
...
+#include "../unzip/zip.h"
+#else
 #include <unzip.h>
 #include <zip.h>
+#endif
 
 namespace ToolKit
 {
+  struct File
```
Comment (afraidofdark) [inline ToolKit/FileManager.h:44]: we don't need this.

File: ToolKit/FullQuadPass.cpp [hunk 50]
```diff
@@ -60,7 +60,7 @@ namespace ToolKit
 
     RenderJobArray jobs;
     RenderJobProcessor::CreateRenderJobs({m_quad}, jobs);
-    renderer->Render(jobs, m_camera, m_params.lights);
+    renderer->Render(jobs, m_camera, {});
```
Comment (afraidofdark) [inline ToolKit/FullQuadPass.cpp:63]: you must also remove the lights from params decleration, this is mislieading.

File: ToolKit/GlErrorReporter.cpp [hunk 51]
```diff
@@ -25,24 +25,24 @@
...
 
 #include "GlErrorReporter.h"
+#include "Logger.h"
+#include "TKOpenGL.h"
 
 namespace ToolKit
 {
 
-  GlReportCallback GlErrorReporter::Report = [](const std::string& msg) -> void { GetLogger()->Log(msg); };
+  GlReportCallback GlErrorReporter::Report = [](const std::string& msg) -> void {  };
```
Comment (afraidofdark) [inline ToolKit/GlErrorReporter.cpp:34]: why did you removed this ? Roll back we need it.

File: ToolKit/RenderSystem.cpp [hunk 52]
```diff
@@ -159,8 +162,9 @@ namespace ToolKit
   void RenderSystem::InitGl(void* glGetProcAddres, GlReportCallback callback)
   {
     // Initialize opengl functions.
+#ifndef __ANDROID__
```
Comment (afraidofdark) [inline ToolKit/RenderSystem.cpp:165]: if def windows instead of ifndef andorid & mscripten

File: ToolKit/Renderer.h [hunk 53]
```diff
@@ -33,6 +33,8 @@
 #include "Types.h"
 #include "Viewport.h"
 
+#define TK_LUT_TEXTURE "GLOBAL_BRDF_LUT_TEXTURE"
```
Comment (afraidofdark) [inline ToolKit/Renderer.h:36]: this should be tk brdf lut tex

File: ToolKit/Renderer.cpp [hunk 54]
```diff
@@ -356,9 +357,9 @@ namespace ToolKit
     }
   }
 
-  void Renderer::SetFramebuffer(FramebufferPtr fb, bool clear, const Vec4& color)
+  void Renderer::SetFramebuffer(FramebufferPtr fb, bool clear, const Vec4& color, bool force)
```
Comment (afraidofdark) [inline ToolKit/Renderer.cpp:360]: create a task to remove these lines after adding the pass that copies render target to back buffer

File: ToolKit/Texture.h (hunk 11 above)
Comment (afraidofdark) [reply ToolKit/Texture.h:33]: this define is duplicate its in renderer.h

File: Dependency/Android/minizip/crypt.h [hunk 55]
```diff
@@ -0,0 +1,131 @@
+/* crypt.h -- base code for crypt/uncrypt ZIPfile
```
Comment (benanil) [inline Dependency/Android/minizip/crypt.h:1]: maybe we can use this with windows and emscripten platform too

File: ToolKit/Renderer.cpp (hunk 54 above)
Comment (benanil) [reply ToolKit/Renderer.cpp:360]: https://oyunteknolojileri.atlassian.net/browse/TOOK-932

File: Dependency/Android/minizip/crypt.h (hunk 55 above)
Comment (benanil) [reply Dependency/Android/minizip/crypt.h:1]: we will use miniz anyway so no need to do this.

File: ToolKit/Util.cpp (hunk 32 above)
Comment (benanil) [reply ToolKit/Util.cpp:484]: probably inadvertandly

File: Editor/main.cpp [hunk 56]
```diff
@@ -112,6 +112,7 @@ namespace ToolKit
         {
           String utf8Path = path.parent_path().u8string();
           utf8Path.erase(remove(utf8Path.begin(), utf8Path.end(), '\"'), utf8Path.end());
+          utf8Path.replace(utf8Path.begin(), utf8Path.end(), '\\', '/');
```
Comment (afraidofdark) [inline Editor/main.cpp:115]: we have unixfy / dosyfiy functions in utils.h which will convert slashes.
why do you convert the correct slash directions ? File system retuns the expected types.
When you need to use path seperator use this util function GetPathSeperator()

Comment (afraidofdark) [reply Editor/main.cpp:115]: Never directly manipulate strings unless they are ascii. All the strings in ToolKit is UTF8 you can't work with it char by char uness the charaters you are working with is in the ascii table.

File: Template/Game.cpp [hunk 57]
```diff
@@ -25,32 +25,117 @@
...
+#include "stdafx.h"
+#include "TKOpenGL.h"
 
 extern "C" TK_GAME_API ToolKit::Game* TK_STDCAL CreateInstance() { return new ToolKit::Game(); }
 
 namespace ToolKit
 {
+  void Game::Init(Main* master)
+  {
+    material = MakeNewPtr<Material>();
```
Comment (afraidofdark) [inline Template/Game.cpp:43]: Don't forget to remove your test codes. This must be a clean empty file.

File: Template/Game.h [hunk 58]
```diff
@@ -1,49 +1,53 @@
...
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+*/
 
 #pragma once
 
-#include <Plugin.h>
-#include <SceneRenderer.h>
-#include <ToolKit.h>
+#include "ToolKit.h"
```
Comment (afraidofdark) [inline Template/Game.h:29]: this is wrong, convert back to <>

File: ToolKit/DofPass.cpp [hunk 59]
```diff
@@ -86,7 +86,6 @@ namespace ToolKit
     m_quadPass->m_params.BlendFunc        = BlendFunction::NONE;
     m_quadPass->m_params.ClearFrameBuffer = false;
     m_quadPass->m_params.FragmentShader   = m_dofShader;
-    m_quadPass->m_params.lights           = {};
```
Comment (afraidofdark) [inline ToolKit/DofPass.cpp:89]: you must remove the lights from the parameters of the quad pass.

File: ToolKit/FileManager.h [hunk 60]
```diff
@@ -28,8 +28,16 @@
 
 #include "Types.h"
 
+#include <variant>
+#include <zlib.h>
+
+#ifdef __ANDROID__
+#include "Android/minizip/unzip.h"
```
Comment (afraidofdark) [inline ToolKit/FileManager.h:35]: use <> any file which is not in the project ( dependencies are in the include list, not in the project )

File: ToolKit/GlErrorReporter.cpp [hunk 61]
```diff
@@ -24,25 +24,24 @@
...
 namespace ToolKit
 {
-
-  GlReportCallback GlErrorReporter::Report = [](const std::string& msg) -> void { GetLogger()->Log(msg); };
+  GlReportCallback GlErrorReporter::Report = [](const String& msg) -> void { GetLogger()->Log(msg); };
 
   void InitGLErrorReport(GlReportCallback callback)
   {
-#ifdef TK_GL_CORE_3_2
+#if defined(TK_GL_CORE_3_2)
```
Comment (afraidofdark) [inline ToolKit/GlErrorReporter.cpp:38]: we used #ifdef everywhere to keep consistency roll this back please.

File: ToolKit/SpriteSheet.cpp [hunk 62]
```diff
@@ -149,9 +149,6 @@ namespace ToolKit
     return true;
   }
 
-  // SpriteAnimation
-  //////////////////////////////////////////////////////////////////////////
-
```
Comment (afraidofdark) [inline ToolKit/SpriteSheet.cpp:154]: put the seperator back please.
And if there are multiple classes that doesnot fit in 150 lines please consider seperating them in the cpp like this.

File: ToolKit/Util.h [hunk 63]
```diff
@@ -221,6 +228,9 @@ namespace ToolKit
     auto it = std::find(arr.cbegin(), arr.cend(), val);
     return it == arr.cend() ? -1 : int(it - arr.cbegin());
   }
+  
+  template<typename T, uint64  N>
+  inline constexpr uint64 ArraySize(const T (&)[N]) { return N; }
```
Comment (afraidofdark) [inline ToolKit/Util.h:233]: please add this when you use it.

**Review by afraidofdark — APPROVED** [review]
